File: backend/proto/community/src/user/domain/user_service.rs

use std::sync::Arc;

use crate::user::domain::user_domain::{Gender, UserDomain};
use crate::user::domain::user_repository::UserRepository;

pub struct UserService {
    repository: Arc<dyn UserRepository>,
}

impl UserService {
    pub fn new(repository: Arc<dyn UserRepository>) -> Self {
        Self { repository }
    }

    pub async fn create_user(&self, email: String, name: String, gender: Gender) -> anyhow::Result<UserDomain> {
        let user = UserDomain::new(email, name, gender);
        self.repository.save(&user).await
    }

    pub async fn get_user_by_id(&self, id: &str) -> anyhow::Result<Option<UserDomain>> {
        self.repository.find_by_id(id).await
    }

    pub async fn get_user_by_email(&self, email: &str) -> anyhow::Result<Option<UserDomain>> {
        self.repository.find_by_email(email).await
    }

    pub async fn update_user(&self, email: &str, updated: UserDomain) -> anyhow::Result<Option<UserDomain>> {
        self.repository.update_by_email(email, &updated).await
    }

    pub async fn list_users(&self) -> anyhow::Result<Vec<UserDomain>> {
        self.repository.find_all().await
    }

    pub async fn delete_user(&self, email: &str) -> anyhow::Result<bool> {
        self.repository.delete_by_email(email).await
    }
}

File: backend/java/community/src/main/java/akiora/user/domain/service/UserService.java
package akiora.user.domain.service;

import lombok.RequiredArgsConstructor;
import akiora.user.domain.model.User;
import akiora.user.mapper.UserMapper;
import akiora.user.persistence.repository.UserRepository;
import org.springframework.stereotype.Service;

import java.util.List;

import static akiora.Constant.LEAGUE_OF_LEGENDS;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final UserMapper mapper;

    public User createUser(User user) {
        var dataUser = mapper.domainToData(user);
        dataUser.setGames(List.of(LEAGUE_OF_LEGENDS));
        return mapper.dataToDomain(userRepository.save(dataUser));
    }

    public User updateUser(User user) {
        checkUserExistsById(user.getId());
        var dataUser = mapper.domainToData(user);
        dataUser.setGames(List.of(LEAGUE_OF_LEGENDS));
        return mapper.dataToDomain(userRepository.save(dataUser));
    }

    public User getUserById(String id) {
        return mapper.dataToDomain(userRepository.findById(id).get());
    }

    public User getUserByEmail(String email) {
        return mapper.dataToDomain(userRepository.findByEmail(email));
    }

    public User deleteUserById(String id) {
        var user = getUserById(id);
        userRepository.deleteById(id);
        return user;
    }

    private void checkUserExistsById(String id) {
        userRepository.findById(id).orElseThrow();
    }
}
